gRPC integration and aio interceptors

fdellekart opened the pull request:
This PR builds on the changes pushed to #2055 . The commits from there were squashed and reworded to adhere to standards described in the contributing guidelines.

Overall, the following functionality was added:
- Client and server interceptors to instrument gRPC calls.
- A gRPC integration to monkeypatch the gRPC functions used for channels and servers.

What's still missing:
- `unary_stream` interceptor for the client, I didn't yet get to looking into that into detail.

fdellekart commented on sentry_sdk/integrations/grpc/aio/client.py:
```diff
@@ -0,0 +1,56 @@
+from sentry_sdk.consts import OP
+
+
+class AsyncClientInterceptor(UnaryUnaryClientInterceptor):
```
Good point, I started writing this in the other module and avoided the name clash

fdellekart commented:
I made the changes you suggested @alisoam , thx for raising :+1: :smile: 

Started writing the unary-stream interceptor for async also, however, I would need some clarification on this topic. I am not extremely familiar with those functionalities and would rather first coordinate.

Is the behavior implemented in the sync version for unary-stream the desired one? It creates one span when the request is incoming which ends as soon as the `UnaryStreamCall` object obtained from the `continuation` is returned. However, this is only half the story, because AFAIU later the client will iterate over that object to obtain the individual messages from the server.

So I think the solution could be to return a custom `UnaryStreamCall` object which wraps also the `read` method of the underlying object into a span. Due to my atm quite limited understanding of the internals of this, this is mostly guesswork, however.

If there's somebody here who can give me a bit more of an insight I can add it to this PR, else I'd suggest opening a follow up issue for the other types of calls (stream-unary and stream-stream were not touched yet at all) :+1: :wink:

antonpirker commented:
Hey! I am looking at this now. One first thing, could you fix the type errors in the failed check `CI / Lint Sources`? (You can run the linting locally with `./scripts/runtox.sh linters`. At least I hope this works in your local environment)

antonpirker commented:
@fdellekart About your question about the "sync version for unary-stream" behavior. I am really not deep into grpc. But I guess for the first version this behavior should be OK. We can always release an update to the integration to improve things. For now I would say we try to bring this out the door.

antonpirker commented:
I am trying to test this and I have a example script resembling this: https://stackoverflow.com/a/63020695/300572

And I get this error:
```
grpc.aio._call.AioRpcError: <AioRpcError of RPC that terminated with:
        status = StatusCode.UNKNOWN
        details = "Unexpected <class 'TypeError'>: object HelloReply can't be used in 'await' expression"
        debug_error_string = "UNKNOWN:Error received from peer  {created_time:"2023-09-27T16:03:03.322818+02:00", grpc_status:2, grpc_message:"Unexpected <class \'TypeError\'>: object HelloReply can\'t be used in \'await\' expression"}"
```

antonpirker commented:
One question @fdellekart or @alisoam 

Previously users had to set the interceptors by hand (see https://docs.sentry.io/platforms/python/configuration/integrations/grpc/) and with this update this is done automatically, if the `GRPCIntegration` is added.

If people add now the new `GRPCIntegration` and forget to remove the "adding interceptors by hand", will this be a problem? Will then be two interceptors do the same thing? Can we make this somehow backwards compatible? 

Of if there is already the Sentry interceptor present, do not add it again (when calling `grpc.intercept_channel()` and in the new `_wrap_*` functions)?

fdellekart commented:
Thank you for the clarifications, I already fixed the linters and added the async interceptor for unary-stream, will add some more tests and then push everything in the upcoming days.

Regarding the error described above, I started the greeter service and client from the official gRPC examples with the integration and I unfortunately can't reproduce your error. What script are you using as the server? As you are getting a response I assume the issue is there somewhere.

Good point regarding the backwards compatibility, I will add some logic to only add the interceptor when not yet present :+1:

antonpirker commented:
I added now a hack to have `ParamSpec` work with older Python versions.

antonpirker commented:
The tests are now running, but the `test_grpc.py::test_grpc_server_starts_transaction()` tests times out and I am not sure why this is.

antonpirker commented:
Hey @fdellekart !
I have now worked a bit on the tests. 

There are still some typing checks failing. (See the "CI / Lint Sources" check above. Can you fix those please?
You can run the check locally by doing (in the repo root directory):
```
pip install -r linter-requirements.txt
mypy sentry_sdk
```

fdellekart commented:
@antonpirker could please explain what we need the extra `.proto` file for `aio` which duplicates the message and service definition of the existing service? AFAIU this isn't necessary and the same generated code can be used for both sync and async versions.

While writing tests for the `unary stream` interceptor I found out that the aio server interceptor works for `unary unary` but actually breaks `unary stream` requests, therefore it's taking a bit longer to get things straight.

Furthermore, I'd suggest to please coordinate who pushes to this branch when, because if we are both working in parallel we end up with conflicts that could be avoided :sweat_smile:

antonpirker commented:
> @antonpirker could please explain what we need the extra .proto file for aio which duplicates the message and service definition 

You are right. They are not necessary. I cleaned them up.

antonpirker commented:
> Furthermore, I'd suggest to please coordinate who pushes to this branch when, because if we are both working in parallel we end up with conflicts that could be avoided 😅

Yes definitely, sorry If I caused you any headaches

I will now will let you work now and will not mess around with the branch. Just ping me if you need anything.

fdellekart commented:
@antonpirker I now fixed the described issues and also added tests to avoid that the integration breaks functionality of currently unsupported RPC types.

The other RPC types apart from unary-unary and unary-stream are now only implemented so that the integration works with them, but there is not yet any tracing supported, whilst the second one is only supported on the client side. I think for now that's enough and from my side, this is good to go to not let this PR grow bigger than it must be.

As soon as I get to it I will open a follow-up PR to implement tracing and error handling for the other RPC types if nobody else does so before me :+1: :smiley: 

AFAIU I cannot trigger the test and linters pipelines myself, so let me know if there's anything I can still help with after those have run :+1:

fdellekart commented:
> Also on the last commit I pushed I found a bug in the sync version of the gRPC `unary stream` client interceptor. It's getting stuck when trying to add the status code to the span. No clue if this worked with older gRPC versions because it was not tested yet.

On reply to that: There's an issue open in grpc on that topic already, for now we can't have the status code on the streaming response because of it.

antonpirker commented:
Thanks for the update! 
I fixed some smaller linting issues. 
I will probably look at this tomorrow, because I am in the middle of something right now.

antonpirker commented:
Sorry for not yet looking at this. I am finishing up something else and then I will look at this. Hopefully this week.

antonpirker commented:
Hey @fdellekart 

Sorry, was caught up with other stuff, but I try to push this over the finish line until tomorrow.

antonpirker commented:
I made a change, to have nicer transaction names in the async server intercepter. @fdellekart can you have a look at the commit: https://github.com/getsentry/sentry-python/pull/2369/commits/7be1306ec3891aa8714cb90f0ac7a7748c7afd39 and tell me if having this _find_name not a static method but a normal method can cause any problems?

antonpirker commented:
Also updated the docs: https://github.com/getsentry/sentry-docs/pull/8434

fdellekart commented:
> I made a change, to have nicer transaction names in the async server intercepter. @fdellekart can you have a look at the commit: [7be1306](https://github.com/getsentry/sentry-python/commit/7be1306ec3891aa8714cb90f0ac7a7748c7afd39) and tell me if having this _find_name not a static method but a normal method can cause any problems?

Shouldn't be a problem :+1:

antonpirker commented:
Nice! So now everything is ready, we on our end just need to fix the problem with the aws tests (they use github secrects and PRs from outside our organization can not access them)

As soon as we have this we can merge and it will be available in the next release (which will be this or next week)!

antonpirker reviewed (approved):
looks very good now! 💯
